File: tauri-app/src-tauri/src/providers/baidu.rs
use crate::providers::models::*;
use crate::providers::Provider;
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;

pub struct BaiduQianfanProvider {
    client: reqwest::Client,
    extra: HashMap<String, String>,
}

impl BaiduQianfanProvider {
    pub fn new(client: reqwest::Client, extra: &HashMap<String, String>) -> Self {
        Self {
            client,
            extra: extra.clone(),
        }
    }

    fn get_cookie(&self) -> Option<&str> {
        self.extra.get("cookie").map(|s| s.as_str()).filter(|s| !s.is_empty())
    }
}

impl Provider for BaiduQianfanProvider {
    fn fetch_usage(&self) -> Pin<Box<dyn Future<Output = UsageData> + Send + '_>> {
        Box::pin(async move {
        let cookie = match self.get_cookie() {
            Some(c) => c,
            None => {
                return UsageData::no_api_key("baidu", "百度千帆");
            }
        };

        let url = "https://console.bce.baidu.com/api/qianfan/charge/codingPlan/resourceList";

        let mut req = self.client
            .get(url)
            .header("Cookie", cookie)
            .header("Accept", "application/json;charset=UTF-8")
            .header("Content-Type", "application/json")
            .header("Referer", "https://console.bce.baidu.com/qianfan/resource/subscribe")
            .header("x-requested-with", "XMLHttpRequest")
            .timeout(std::time::Duration::from_secs(15));

        if let Some(csrf) = self.extra.get("csrftoken") {
            req = req.header("csrftoken", csrf);
        }
        if let Some(jt) = self.extra.get("x-bce-jt") {
            req = req.header("x-bce-jt", jt);
        }

        let resp = match req.send().await {
            Ok(r) => r,
            Err(e) => return UsageData::error("baidu", "百度千帆", &e.to_string()),
        };

        if resp.status() == reqwest::StatusCode::UNAUTHORIZED
            || resp.status() == reqwest::StatusCode::FORBIDDEN
        {
            return UsageData::unauthorized("baidu", "百度千帆", "Cookie 已失效，请重新登录百度智能云");
        }

        let data: serde_json::Value = match resp.json().await {
            Ok(d) => d,
            Err(e) => return UsageData::error("baidu", "百度千帆", &e.to_string()),
        };

        // Check for login redirect
        if data.get("success").and_then(|v| v.as_bool()) == Some(false) {
            if let Some(msg) = data.get("message").and_then(|v| v.as_object()) {
                if msg.get("redirect").is_some() {
                    return UsageData::unauthorized(
                        "baidu",
                        "百度千帆",
                        "Cookie 已失效，请重新登录百度智能云",
                    );
                }
            }
        }

        // Parse response
        if data.get("success").and_then(|v| v.as_bool()) != Some(true) {
            let msg = data
                .get("message")
                .and_then(|v| v.as_str())
                .unwrap_or("请求失败");
            return UsageData::error("baidu", "百度千帆", msg);
        }

        let items = data
            .get("result")
            .and_then(|r| r.get("items"))
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        if items.is_empty() {
            return UsageData::error("baidu", "百度千帆", "未找到用量数据");
        }

        let item = &items[0];
        let plan_name = item
            .get("planType")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();

        let quota = item.get("quota").cloned().unwrap_or(serde_json::json!({}));
        let mut windows: Vec<UsageWindow> = Vec::new();

        let window_defs = vec![
            ("fiveHour", "5小时配额"),
            ("week", "每周配额"),
            ("month", "每月配额"),
        ];

        for (key, label) in window_defs {
            let entry = match quota.get(key) {
                Some(e) => e,
                None => continue,
            };

            let used = entry.get("used").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let limit = entry.get("limit").and_then(|v| v.as_f64()).unwrap_or(0.0);
            if limit <= 0.0 {
                continue;
            }

            let resets_at = entry
                .get("resetAt")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());

            windows.push(UsageWindow {
                label: label.into(),
                used_percent: ((used / limit) * 100.0).min(100.0),
                used,
                total: limit,
                remaining: limit - used,
                resets_at,
                unit: "次".into(),
            });
        }

        let is_empty = windows.is_empty();
        UsageData {
            provider_id: "baidu".into(),
            provider_name: "百度千帆".into(),
            status: if is_empty {
                ProviderStatus::Error
            } else {
                ProviderStatus::Ok
            },
            plan_name,
            windows,
            balance: None,
            error_message: if is_empty {
                "无法解析用量数据".into()
            } else {
                String::new()
            },
            updated_at: UsageData::now_timestamp(),
        }
        })
    }
}

File: src/providers/baidu.py
"""Baidu Qianfan (百度千帆) provider - Cookie-based authentication."""

from __future__ import annotations

from datetime import datetime

import httpx

from .base import BaseProvider, ProviderStatus, UsageData, UsageWindow


class BaiduQianfanProvider(BaseProvider):
    """百度千帆编程套餐用量 Provider（Cookie 认证）。

    用户通过在浏览器中复制 curl 命令来配置认证信息。
    """

    API_URL = "https://console.bce.baidu.com/api/qianfan/charge/codingPlan/resourceList"

    def __init__(self, api_key: str = "", **kwargs):
        super().__init__(
            provider_id="baidu",
            name="百度千帆",
            description="百度千帆编程套餐用量",
            requires_api_key=False,
            api_key=api_key,
            **kwargs,
        )

    def get_api_key(self) -> str | None:
        return None

    def _get_cookie_header(self) -> str | None:
        manual = self.extra_config.get("cookie", "")
        if manual:
            return manual
        return None

    def _get_auth_headers(self) -> dict[str, str]:
        """从 extra_config 获取额外的认证请求头。"""
        headers: dict[str, str] = {}
        csrftoken = self.extra_config.get("csrftoken", "")
        if csrftoken:
            headers["csrftoken"] = csrftoken
        x_bce_jt = self.extra_config.get("x-bce-jt", "")
        if x_bce_jt:
            headers["x-bce-jt"] = x_bce_jt
        return headers

    async def fetch_usage(self) -> UsageData:
        cookie = self._get_cookie_header()
        if not cookie:
            return UsageData(
                provider_id=self.provider_id,
                provider_name=self.name,
                status=ProviderStatus.NO_API_KEY,
                error_message="请先在设置中粘贴 curl 命令",
            )

        headers = {
            "Cookie": cookie,
            "Accept": "application/json;charset=UTF-8",
            "Content-Type": "application/json",
            "Referer": "https://console.bce.baidu.com/qianfan/resource/subscribe",
            "x-requested-with": "XMLHttpRequest",
        }
        headers.update(self._get_auth_headers())

        try:
            async with httpx.AsyncClient(timeout=15, trust_env=False) as client:
                resp = await client.get(self.API_URL, headers=headers)

                if resp.status_code in (401, 403):
                    return UsageData(
                        provider_id=self.provider_id,
                        provider_name=self.name,
                        status=ProviderStatus.UNAUTHORIZED,
                        error_message="Cookie 已失效，请重新登录百度智能云",
                    )
                resp.raise_for_status()
                data = resp.json()

        except httpx.HTTPError as e:
            return UsageData(
                provider_id=self.provider_id,
                provider_name=self.name,
                status=ProviderStatus.ERROR,
                error_message=str(e),
            )

        # 检查登录重定向
        if not data.get("success") and isinstance(data.get("message"), dict):
            if data["message"].get("redirect"):
                return UsageData(
                    provider_id=self.provider_id,
                    provider_name=self.name,
                    status=ProviderStatus.UNAUTHORIZED,
                    error_message="Cookie 已失效，请重新登录百度智能云",
                    raw_response=data,
                )

        return self._parse_response(data)

    def _parse_response(self, data: dict) -> UsageData:
        if not data.get("success"):
            return UsageData(
                provider_id=self.provider_id,
                provider_name=self.name,
                status=ProviderStatus.ERROR,
                error_message=data.get("message", "请求失败") if isinstance(data.get("message"), str) else "请求失败",
                raw_response=data,
            )

        items = data.get("result", {}).get("items", [])
        if not items:
            return UsageData(
                provider_id=self.provider_id,
                provider_name=self.name,
                status=ProviderStatus.ERROR,
                error_message="未找到用量数据",
                raw_response=data,
            )

        item = items[0]
        plan_name = item.get("planType", "")
        quota = item.get("quota", {})

        windows: list[UsageWindow] = []
        window_defs = [
            ("fiveHour", "5小时配额"),
            ("week", "每周配额"),
            ("month", "每月配额"),
        ]

        for key, label in window_defs:
            entry = quota.get(key)
            if not entry:
                continue

            used = float(entry.get("used", 0))
            limit = float(entry.get("limit", 0))
            if limit <= 0:
                continue

            resets_at = None
            reset_str = entry.get("resetAt", "")
            if reset_str:
                try:
                    resets_at = datetime.fromisoformat(reset_str)
                except (ValueError, TypeError):
                    pass

            windows.append(
                UsageWindow(
                    label=label,
                    used_percent=min((used / limit) * 100, 100),
                    used=used,
                    total=limit,
                    remaining=limit - used,
                    resets_at=resets_at,
                    unit="次",
                )
            )

        return UsageData(
            provider_id=self.provider_id,
            provider_name=self.name,
            status=ProviderStatus.OK if windows else ProviderStatus.ERROR,
            plan_name=plan_name,
            windows=windows,
            error_message="" if windows else "无法解析用量数据",
            raw_response=data,
        )

    def get_display_config(self) -> dict[str, str]:
        return {"curl": "textarea"}

    @classmethod
    def parse_curl(cls, curl_text: str) -> dict[str, str]:
        """解析 curl 命令文本，返回提取的认证字段。"""
        from src.utils.curl_parser import parse_curl

        return parse_curl(curl_text)
